### PR TITLE
Fix reentrancy with WidgetBindingObserver callbacks

### PR DESCRIPTION
**Pull request by dnfield**
Part of https://github.com/flutter/flutter/issues/131678

Fixes up callsites for WidgetsBindingObserver related callbacks.

### CODE REVIEW

**Comment by gspencergoog** [comment]
Looks like this might cause some observers to not be called (at least `didPushRoute` and `didPushRouteInformation`, from the presubmit failures).

**Comment by dnfield** [comment]
Needed to flush the asynchronous callbacks before removing - otherwise the new observer was sticking around after the `removeSelf` call and interfering with other tests.

File: packages/flutter/test/widgets/binding_test.dart [hunk 1]
```diff
@@ -141,7 +155,9 @@ void main() {
     WidgetsBinding.instance.handlePopRoute();
     WidgetsBinding.instance.handlePushRoute('/');
     WidgetsBinding.instance.handleRequestAppExit();
-    expect(observer.removeSelf(), greaterThan(11));
+    await tester.idle();
+    expect(observer.removeSelf(), greaterThan(1));
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/binding_test.dart:159]: Did you mean to change this from 11 to 1?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/test/widgets/binding_test.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/test/widgets/binding_test.dart:159]: I don't want to use an exact value cause this goes up or down pretty quickly depending on how many methods get called. I think it's probably ok to just make sure it got added more than in the constructor, so I put it to 1.

**Pull request merged**